### ISSUE TITLE
vcf2bed and deletions

### ISSUE THREAD

**Issue by nathanhaigh** [bedops/bedops#218]
When converting vcf to BED I think the script should probably deal with deletions better. The resulting BED range should define the size of the deletion rather than just the single base preceding the deletion.
My understanding is that when `ALT = <DEL>` the length of the deletion should be specified in SVLEN attribute.

I came across this issue when trying to use BEDOPS to remove deletion variants from a VCF file which overlapped the assembly gaps specified in a GFF3 file:

```bash
bedops -n 1  \
  <(pigz -dcp2 my.del.vcf.gz | vcf2bed) \
  <(pigz -dcp2 my.gaps.gff3.gz | gff2bed)
```

**Comment by alexpreynolds** [comment]
If you use `vcf2bed --deletions` to convert VCF deletions to BED, I believe that the stop value of the BED output is determined by the length difference between `ALT` and `REF` alleles. This appears to be equivalent to the definition of the `SVLEN` attribute (or should be identical, unless there is a bug).

Changing the default behavior of `vcf2bed` when not used with the `--deletions` option may be useful, but I think I kept this behavior specific to the `--deletions` option to preserve how older versions of this conversion tool once worked. 

I'd appreciate any feedback on this, or if `--deletions` does not behave as expected.

Cf. https://bedops.readthedocs.io/en/latest/content/reference/file-management/conversion/vcf2bed.html

**Comment by nathanhaigh** [comment]
Thanks @alexpreynolds for the prompt response. Adding the `--deletions` option doesn't work, it generates zero output. :(

My VCF file is generated by Sniffles and deletions look like this:

```
chr1H   270744  0       N       <DEL>   .       PASS    PRECISE;SVMETHOD=Snifflesv1.0.10;CHR2=chr1H;END=271039;ZMW=13;STD_quant_start=0.000000;STD_quant_stop=1.449138;Kurtosis_quant_start=2.000000;Kurtosis_quant_stop=-0.910583;SVTYPE=DEL;SUPTYPE=AL,SR;SVLEN=-295;STRANDS=+-;RE=14;REF_strand=7,8;AF=0.482759      GT:DR:DV        0/1:15:14
chr1H   1149919 2       N       <DEL>   .       PASS    PRECISE;SVMETHOD=Snifflesv1.0.10;CHR2=chr1H;END=1150194;ZMW=12;STD_quant_start=4.593474;STD_quant_stop=5.718391;Kurtosis_quant_start=-0.103263;Kurtosis_quant_stop=-1.399870;SVTYPE=DEL;SUPTYPE=AL,SR;SVLEN=-275;STRANDS=+-;RE=13;REF_strand=3,7;AF=0.565217    GT:DR:DV        0/1:10:13
```

**Comment by alexpreynolds** [comment]
Thanks for the heads up. I'll need to take a look at the v4.3 specification and see what changes are needed to deal with how these are formatted. I'll leave this issue open until I get an update in place.

**Comment by nathanhaigh** [comment]
As a work around I'm doing some post-processing with `gawk` to use the `SVLEN` to munge the start and end coordinates of the BED output:

```bash
pigz -dcp2 merged/merged.genotype.del.vcf.gz \
  | vcf2bed \
  | gawk '{match($9, /SVLEN=-?([^\;]*)/, a); $3+=a[1]+1; print}'
```
Let me know if you think I have any off-by-1 errors as I have only had a quick check of this and generated no test cases.

**Comment by alexpreynolds** [comment]
I think that VCF uses a one-based index. To convert a single-based position to a zero-based, half-open start position, I subtract one from the `POS` value. The stop position is the start position, plus one (with the exception of deletions).

https://github.com/bedops/bedops/blob/master/applications/bed/conversion/src/convert2bed.c#L5078-L5079

When making a zero-based BED file, the difference of the start and stop coordinates should equal SVLEN.

With that in mind, if I understand the definition of `SVLEN` correctly, it should be enough to add the value of `SVLEN` to the zero-based start position that comes out of `vcf2bed`. 

In other words, I don't think that you need to add one to `$3`, and you would just use `$3 += a[1]`. 

One way to test this is to take the absolute difference of start and stop coordinates and compare with `SVLEN` in the final product.

**Comment by alexpreynolds** [comment]
In fact, I think you would use `$3 += a[1] - 1`, as `$3` (the stop coordinate) was already incremented by `vcf2bed`.

Taking the difference of `$2` and `$3` and comparing against `SVLEN` should confirm what to do.

**Comment by nathanhaigh** [comment]
You are correct @alexpreynolds, my end coordinate was off-by-one (+1).

Since the `END` position of the deletion is also reported, I'm simply extracting that and using it instead of doing the arithmetic with `SVLEN`. Although, I'm not sure this is a general approach and whether `END` will always be set by other SV callers.

```bash
pigz -dcp2 merged/merged.genotype.del.vcf.gz \
  | vcf2bed \
  | gawk '{match($9, /END=([^\;]*)/, a); $3=a[1]; print}'
```

**Comment by johnemajor** [comment]
This was a huge help!  Thanks Nathan.

**Comment by gungorbudak** [comment]
Is this difference with `vcf2bed` (v2.4.35) expected when converting deletions? 

```
bgzip -cd gnomad.genomes.r3.0.sites.chr21.af001.vcf.gz | cut -f1-7 | grep 'rs1239803462'
chr21	5031125	rs1239803462	CT	C	1.99592e+06	PASS
bgzip -cd gnomad.genomes.r3.0.sites.chr21.af001.vcf.gz | vcf2bed - | cut -f1-7 | grep 'rs1239803462'
chr21	5031124	5031125	rs1239803462	1.99592e+06	CT	C
bgzip -cd gnomad.genomes.r3.0.sites.chr21.af001.vcf.gz | vcf2bed -n - | cut -f1-7 | grep 'rs1239803462'
chr21	5031124	5031126	rs1239803462	1.99592e+06	CT	C
```

The option `-n` returns the correct start and end I guess and it looks like I should convert snps/insertions/deletions separately.

**Comment by alexpreynolds** [comment]
Yes, this is expected:

> _[By default, the `vcf2bed` script translates all variants to single-base positions in the resulting BED output.](https://bedops.readthedocs.io/en/latest/content/reference/file-management/conversion/vcf2bed.html#customized-variant-handling) Depending on the category of variant you are interested in, however, you may want more specific categories handled differently..._
>
> _When using `--deletions`, the stop value of the BED output is determined by the length difference between ALT and REF alleles. Use of `--insertions` or `--snvs` yields a one-base BED element._

I suggest using the explicit options to get specific variant categories.
